jlm/hls/backend/rhls2firrtl/verilator-harness-hls.cpp

```diff
@@ -1,743 +1,645 @@
   {
-    cpp << "    top->i_data_" << i << " = 0;\n";
-  }
-  cpp << "    top->reset = 1;\n"
-         "\n";
-  for (size_t i = 0; i < mem_reqs.size(); ++i)
-  {
-    cpp << "    top->mem_" << i
-        << "_req_ready = false;\n"
-           "    top->mem_"
-        << i
-        << "_res_valid = false;\n"
-           "    top->mem_"
-        << i << "_res_data_data = 0xDEADBEEF;\n";
+    return "int" + util::strfmt(t->nbits()) + "_t";
   }
-  cpp << "    clock_cycle();\n"
-         "    clock_cycle();\n"
-         "    top->reset = 0;\n"
-         "    clock_cycle();\n"
-         "}\n"
-         "\n"
-         "void posedge() {\n"
-         "    if (terminate) {\n"
-         "        std::cout << \"terminating\\n\";\n"
-         "        verilator_finish();\n"
-         "        exit(-1);\n"
-         "    }\n"
-         "    assert(!Verilated::gotFinish());\n"
-         "    top->clk = 1;\n"
-         "    top->eval(); //eval here to get a clean posedge with the old inputs\n"
-         "}\n"
-         "\n"
-         "void finish_clock_cycle() {\n";
-  for (size_t i = 0; i < mem_reqs.size(); ++i)
+  if (jlm::rvsdg::is<llvm::PointerType>(*type))
   {
-    cpp << "    top->mem_" << i << "_res_data_data = mem_resp[" << i
-        << "]->front().data;\n"
-           "    top->mem_"
-        << i << "_res_data_id = mem_resp[" << i
-        << "]->front().id;\n"
-           "    top->mem_"
-        << i << "_res_valid = mem_resp[" << i
-        << "]->front().valid;\n"
-           "    top->mem_"
-        << i << "_req_ready = true;\n";
+    return "void*";
   }
-  cpp << "    top->eval();\n"
-         "    // dump before trying to access memory\n"
-         "#ifdef TRACE_SIGNALS\n"
-         "    tfp->dump(main_time * 2);\n"
-         "#ifdef VCD_FLUSH\n"
-         "    tfp->flush();\n"
-         "#endif\n"
-         "#endif\n";
-  for (size_t i = 0; i < mem_reqs.size(); ++i)
+  if (auto t = dynamic_cast<const llvm::arraytype *>(type))
   {
-    cpp << "    if (top->mem_" << i << "_res_valid && top->mem_" << i
-        << "_res_ready) {\n"
-           "        mem_resp["
-        << i
-        << "]->pop();\n"
-           "    } else if (!mem_resp["
-        << i
-        << "]->front().valid){\n"
-           "        mem_resp["
-        << i
-        << "]->pop();\n"
-           "    }\n";
+    return convert_to_c_type(&t->element_type()) + "*";
   }
-  for (size_t i = 0; i < mem_reqs.size(); ++i)
+
+  JLM_UNREACHABLE("Unimplemented C type");
+}
+
+/**
+ * Takes an HLS kernel and determines the return type of the original C function.
+ * Returns a tuple where the first element is a boolean signifying if anything is returned.
+ * The second element is a string containing the kernel's return type when called from C.
+ * @param hls_lambda the lambda node representing the kernel
+ * @return a tuple (has return value, return type)
+ */
+std::pair<bool, std::string>
+convert_to_c_return_type(const llvm::lambda::node & hls_lambda)
```

Comment:
Naming convention of function and argument.
hlsLambda or should we simply start calling it the “kernel”?

Reply:
Changed